### PR TITLE
feat: migrate image to distroless

### PR DESCRIPTION
**Pull request by Demonsthere**
## Checklist

- [ ] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [ ] I have referenced an issue containing the design document if my change
      introduces a new feature.
- [ ] I am following the
      [contributing code guidelines](../blob/master/CONTRIBUTING.md#contributing-code).
- [ ] I have read the [security policy](../security/policy).
- [ ] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security vulnerability, I
      confirm that I got the approval (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [ ] I have added tests that prove my fix is effective or that my feature
      works.
- [ ] I have added or changed [the documentation](https://github.com/ory/docs).

## Further Comments
- ~~Downgrade alpine image to 3.17 due to [failing package upgrade](https://github.com/ory/hydra/actions/runs/5188782643/jobs/9352884594#step:8:110)~~
- Migrate to debian builder and distroless:base image to avoid CVE issues in base containers. We cannot migrate to static as hydra is build with dynamic linking.

### CODE REVIEW

File: .docker/Dockerfile-build [hunk 1]
```diff
@@ -17,28 +18,14 @@ COPY . .
 
 RUN go build -tags sqlite,json1 -o /usr/bin/hydra
 
-FROM alpine:3.18
+#########################
 
-RUN addgroup -S ory; \
-    adduser -S ory -G ory -D  -h /home/ory -s /bin/nologin; \
-    chown -R ory:ory /home/ory
+FROM gcr.io/distroless/static-debian11:nonroot as runner
```
Comment (zepatrik) [inline .docker/Dockerfile-build:23]: Did we check whether this has a compatible glibc, i.e. works with SQLite? Note that here we don't use musl, so the binary has to be linked against glibc.

**Review by zepatrik — COMMENTED** [review]
If it works with sqlite :+1:

File: .docker/Dockerfile-build (hunk 1 above)
Comment (hperl) [reply .docker/Dockerfile-build:23]: From what I remember you have to set `CGO_ENABLED 0` for Go to work in non-libc environments.

File: .docker/Dockerfile-build [hunk 2]
```diff
@@ -1,44 +1,28 @@
...
 ENV GO111MODULE on
 ENV CGO_ENABLED 1
 
-RUN go mod download
-
 COPY . .
+RUN go mod download &&\
+    go build -tags sqlite,json1 -o /usr/bin/hydra
 
-RUN go build -tags sqlite,json1 -o /usr/bin/hydra
```
Comment (aeneasr) [inline .docker/Dockerfile-build:18]: Please keep these instructions as they were - they prevent excessive downloading of go modules. Thanks!

**Review by aeneasr — DISMISSED** [review]

**Review by zepatrik — APPROVED** [review]

**Review by zepatrik — APPROVED** [review]

**Pull request merged**